src/HOCON/Impl/HoconTokenizer.cs

```diff
@@ -152,62 +65,113 @@ public string Take(int length)
         /// </summary>
         /// <param name="patterns">The string array that contains the characters to match.</param>
         /// <returns><c>true</c> if any one of the patterns match, otherwise <c>false</c>.</returns>
-        public bool Matches(params string[] patterns)
+        protected bool Matches(params string[] patterns)
         {
             foreach (string pattern in patterns)
             {
-                if (pattern.Length + _index >= _text.Length)
+                if (pattern.Length + Index > _text.Length)
                     continue;
 
-                if (_text.Substring(_index, pattern.Length) == pattern)
+                if (_text.Substring(Index, pattern.Length) == pattern)
                     return true;
             }
             return false;
         }
 
-        /// <summary>
-        /// Retrieves the next character in the tokenizer without advancing its position.
-        /// </summary>
-        /// <returns>The character at the tokenizer's current position.</returns>
-        public char Peek()
+        protected bool Matches(char pattern)
         {
             if (EoF)
-                return (char) 0;
+                return false;
+
+            return _text[Index] == pattern;
+        }
+
+        protected bool Matches(params char[] patterns)
+        {
+            if (EoF)
+                return false;
+
+            return patterns.Any(pattern => _text[Index] == pattern);
+        }
+
+        protected bool BackMatch(string pattern)
+        {
+            if (Index - pattern.Length < 0)
+                return false;
 
-            return _text[_index];
+            return _text.Substring(Index = pattern.Length, pattern.Length) == pattern;
```

Comment:
I don't know which version of .NET we want to target, but I'm pretty sure we can do this check without allocating.